### PR TITLE
[WIP] Update for RidePATH changes and implement SignalR receiver

### PR DESCRIPTION
**Pull request by wgraham17**
Still need to fix and complete unit tests. This addresses #4

### CODE REVIEW

**Comment by mrazza** [comment]
#4 

Thanks! Fix tests/improve test coverage.

**Comment by wgraham17** [comment]
> 
> 
> #4
> 
> Thanks! Fix tests/improve test coverage.

👀

**Comment by mrazza** [comment]
We should also solve for the situation when the JWS tokens for the SignalR hub expire.

**Comment by wgraham17** [comment]
> 
> 
> We should also solve for the situation when the JWT tokens for the SignalR hub expire.

We _probably_ only care about the JWT token expiring when the connection to the hub is established; however, this raises a good point that we don't currently recover from disconnections. We need some sort of recovery mechanism if the SignalR hub connection is dropped to retrieve a new token and reconnect.

**Comment by wgraham17** [comment]
RE: coverage and unit tests for SignalR, SignalR Core is pretty shitty at the moment and it seems like we'll have to create our own `Hub` implementation and hook it into a full blown ASP.NET Core `TestHost` (instead of mocking things on `HubConnection` as it doesn't implement an interface). Maybe we can look at the internals of `HubConnection` and stub out fake protocol/connection handlers?

**Comment by mrazza** [comment]
I've deployed this change to production and will monitor for stability.

Interesting concerns w/r/t testing this. Hard to justify significant investment given the fact that the underlying server contract could change at anytime and we'd just be testing our understanding of it. Let's think a bit more about this or experiment with some of the possible approaches you've outlined before giving up.

File: server/PathServices/SignalRRealtimeDataRepository.cs [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+                {
+                    if (e != null)
+                    {
+                        Log.Logger.Here().Warning(e, "SignalR connection was closed (unexpectedly?) as a result of an exception; recovering");
+                        await Task.Delay(new Random().Next(0, 50) * 100);
+                        await this.CreateHubConnection(tokenBrokerUrl, tokenValue, station, direction);
+                    }
+                    else
+                    {
+                        Log.Logger.Here().Warning("SignalR connection close event was triggered but the exception is null; expected?");
```
Comment (mrazza) [inline server/PathServices/SignalRRealtimeDataRepository.cs]: This is happening for no reason,

In other words, the connection can be terminated without a valid exception -- potentially from the remote host or due to network issues.

https://i.imgur.com/QrouLsw.png

Seems like we should trigger a new connection in this case. We should test triggering a reconnect ourselves to see if Closed is called when it is the client that closes the connection.

File: server.Tests/PathServices/SignalRRealtimeDataRepositoryTest.cs [hunk 2]
```diff
@@ -0,0 +1,10 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace PathApi.Server.Tests.PathServices
+{
+    class SignalRRealtimeDataRepositoryTest
```
Comment (mrazza) [inline server.Tests/PathServices/SignalRRealtimeDataRepositoryTest.cs]: Can you add a TODO() for this test coverage to be implemented?

We can do this in a follow-up.

File: server/PathServices/SignalRRealtimeDataRepository.cs [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+        }
+
+        /// <summary>
+        /// Gets the latest realtime train arrival data for the specified station.
+        /// </summary>
+        /// <param name="station">The station to get realtime arrival data for.</param>
+        /// <returns>A collection of arriving trains.</returns>
+        public Task<IEnumerable<RealtimeData>> GetRealtimeData(Station station)
+        {
+            return Task.FromResult(this.GetRealtimeData(station, RouteDirection.ToNY).Union(this.GetRealtimeData(station, RouteDirection.ToNJ)).Where(data => data.DataExpiration > DateTime.UtcNow));
```
Comment (mrazza) [inline server/PathServices/SignalRRealtimeDataRepository.cs]: I think you're right that Concat is better than Union here as we don't need it to check for duplicates and deal with deduping. (IIUC Union will remove duplicates)

File: server/PathServices/SignalRRealtimeDataRepository.cs [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+                        return;
+                    }
+
+                    if (e != null)
+                    {
+                        Log.Logger.Here().Warning(e, "SignalR connection was closed as a result of an exception");
+                    }
+
+                    Log.Logger.Here().Information("Recovering SignalR connection to {station}-{direction}...", station, direction);
+                    await RetryConnection();
```
Comment (mrazza) [inline server/PathServices/SignalRRealtimeDataRepository.cs]: Should we short-circuit out of here whenever we need to RetryConnection? (both here and in the catch block below)

**Pull request merged**